src/cfa_config_generator/utils/epinow2/functions.py

```diff
@@ -86,8 +91,98 @@ def generate_default_job_id(as_of_date: str | None = None) -> str:
+    report_date: date | None = None,
+    reference_dates: list[date] | None = None,
+    data_source: str | None = None,
+    data_path: str | None = None,
+    data_container: str | None = None,
+    production_date: date | None = None,
+    job_id: str | None = None,
+    as_of_date: str | None = None,
+) -> str:
+    """
+    Confirms that file exists at the path listed, within the given data container,
+    and with the required variables state, disease, reference_date, report_date
+    """
+
+    class CustomError(Exception):
+        pass
+
+    excl_df = pl.read_csv(exclusions)
+
+    df_cols = ["state", "disease", "report_date", "reference_date"]
+    excl_cols = set(excl_df.columns)
+    missing_cols = list(set(df_cols).difference(excl_cols))
+    if len(missing_cols) > 0:
+        raise CustomError(f"data exclusions file missing: {missing_cols}")
+
+    args_dict = {}
+    args_dict["state"] = excl_df.get_column("state").to_list()
+    args_dict["disease"] = excl_df.get_column("disease").to_list()
+    # args_dict["report_date"] = excl_df.get_column("report_date").to_list()
+    # args_dict["reference_date"] = excl_df.get_column("reference_date").to_list()
+    args_dict["exclusions"] = exclusions
+
+    return args_dict
+
+
+def generate_tasks_excl_from_data_excl(excl_df: pl.DataFrame) -> str:
+    """
+    Confirms that file exists at the path listed, within the given data container,
+    and with the required variables state, disease, reference_date, report_date.
+    Next, creates an output string in the task exclusion form in the state:disease
+    pair form with all of the states and disease to not have tasks created for.
+
+    Parameters
+    -----------
+        excl_df: a dataframe of the exclusions
+
+    Returns
+    -----------
+        A string of the form state:disease,state:disease
+        for all of the state:disease pairs that should
+        be excluded from the task generation.
+
+    Raises
+    -----------
+        ValueError: If the exclusions file is missing required columns.
+
+    Notes
+    -----------
+    Assumes that `excl_df` has schema
+    [
+        (state, pl.String),
+        (disease, pl.String),
+        (report_date, pl.Date),
+        (reference_date, pl.Date)
+    ]
+    """
+    want_cols: set[str] = {"state", "disease", "report_date", "reference_date"}
+    got_cols: set[str] = set(excl_df.columns)
+    missing_cols: set[str] = want_cols.difference(got_cols)
+    if any(missing_cols):
+        raise ValueError(f"data exclusions file missing: {missing_cols}")
+
+    all_set: set[tuple[str, str]] = set(itertools.product(all_states, all_diseases))
+
+    incl_states = excl_df.get_column("state").to_list()
+    incl_disease = excl_df.get_column("disease").to_list()
+    incl_set: set[tuple[str, str]] = set(zip(incl_states, incl_disease))
+
+    excl_set: set[tuple[str, str]] = all_set.difference(incl_set)
+
+    intermediate_list: list[str] = [es[0] + ":" + es[1] for es in excl_set]
```

Comment:
This is a bit confusing to read -- can we add a comment on what this `intermediate_list` looks like?

Reply:
[added](https://github.com/CDCgov/cfa-config-generator/pull/48/commits/4e0f00f0886fc5effa9e2477c8705518b2003662)